gcc/rust/backend/rust-compile-intrinsic.cc

```diff
@@ -837,6 +929,118 @@ atomic_load_handler_inner (Context *ctx, TyTy::FnType *fntype, int ordering)
+
+  std::vector<Bvariable *> param_vars;
+  std::vector<tree> types;
+  compile_fn_params (ctx, fntype, fndecl, &param_vars, &types);
+
+  if (!ctx->get_backend ()->function_set_parameters (fndecl, param_vars))
+    return error_mark_node;
+
+  enter_intrinsic_block (ctx, fndecl);
+
+  auto dst = ctx->get_backend ()->var_expression (param_vars[0], Location ());
+  auto old = ctx->get_backend ()->var_expression (param_vars[1], Location ());
+  auto src = ctx->get_backend ()->var_expression (param_vars[2], Location ());
+  auto success_memorder = make_unsigned_long_tree (ctx, success_memorder_raw);
+  auto failure_memorder = make_unsigned_long_tree (ctx, failure_memorder_raw);
+
+  TREE_READONLY (dst) = 0;
+
+  auto monomorphized_type
+    = fntype->get_substs ()[0].get_param_ty ()->resolve ();
+  auto builtin_name
+    = build_atomic_builtin_name ("atomic_compare_exchange_",
+				 fntype->get_locus (), monomorphized_type);
+
+  auto atomic_cmp_xchg_raw = NULL_TREE;
+  BuiltinsContext::get ().lookup_simple_builtin (builtin_name,
+						 &atomic_cmp_xchg_raw);
+  rust_assert (atomic_cmp_xchg_raw);
+
+  auto atomic_cmp_xchg = build_fold_addr_expr_loc (Location ().gcc_location (),
+						   atomic_cmp_xchg_raw);
+
+  // The gcc builtin has a fun behavior where if *ptr and *expected are not
+  // equal, it will write *ptr into *expected. So we actually need to create a
+  // tmp variable which is copy(old) and pass the adress as an argument to the
+  // builtin. We then perform the call, and return {tmp, result} as the old
+  // value will have been written into tmp
+
+  // let tmp = old;
+  // let tmp_addr = &tmp;
+  auto tmp_init = NULL_TREE;
+  auto tmp = ctx->get_backend ()
+	       ->temporary_variable (fndecl, NULL_TREE, TREE_TYPE (types[0]),
+				     old, true, Location (), &tmp_init)
+	       ->get_decl ();
+  auto tmp_addr = build_fold_addr_expr (tmp);
+
+  TREE_READONLY (tmp) = 0;
+  TREE_SIDE_EFFECTS (tmp) = 1;
+  TREE_READONLY (tmp_init) = 0;
+  TREE_SIDE_EFFECTS (tmp_init) = 1;
+
+  // let call_res = atomic_compare_exchange(dst, tmp_addr, src, ...);
+  auto cxchg_call = ctx->get_backend ()->call_expression (
+    atomic_cmp_xchg,
+    {dst, tmp_addr, src, weak_value, success_memorder, failure_memorder},
+    nullptr, Location ());
+
+  TREE_READONLY (cxchg_call) = 0;
+  TREE_SIDE_EFFECTS (cxchg_call) = 1;
+
+  auto call = NULL_TREE;
+  auto call_res = ctx->get_backend ()
+		    ->temporary_variable (fndecl, NULL_TREE, boolean_type_node,
+					  cxchg_call, false, Location (), &call)
+		    ->get_decl ();
+
+  TREE_READONLY (call) = 0;
+  TREE_SIDE_EFFECTS (call) = 1;
+
+  // Add both statements to the current function declaration
+  ctx->add_statement (tmp_init);
+  ctx->add_statement (call);
+
+  auto tuple_type = ctx->get_backend ()->struct_type ({
+    Backend::typed_identifier ("__0", TREE_TYPE (types[0]), Location ()),
+    Backend::typed_identifier ("__1", boolean_type_node, Location ()),
+  });
```

Comment:
You should be able to get this from the fndecl type. I think if you simply get TREE_TYPE(fndecl) it will give you this type canonically so you avoid the non trivial conversion issue.

Reply:
a simpler way might even be to use ```TREE_TYPE(DECL_RESULT(fndecl))```

Reply:
The problem mentioned in the test is still present, but this does look much cleaner :) Thanks